src/Dhall/Map.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+-- | `Map` type used to represent records and unions
+
+module Dhall.Map
+    ( -- * Type
+      Map
+
+      -- * Construction
+    , singleton
+    , fromList
+
+      -- * Sorting
+    , sort
+    , isSorted
+
+      -- * Insertion
+    , insert
+    , insertWith
+
+      -- * Deletion/Update
+    , delete
+
+      -- * Query
+    , lookup
+    , member
+
+      -- * Combine
+    , union
+    , unionWith
+    , intersection
+    , intersectionWith
+
+      -- * Traversals
+    , mapWithKey
+    , traverseWithKey
+    , foldMapWithKey
+
+      -- * Conversions
+    , toList
+    , toMap
+    , keys
+    ) where
+
+import Control.Applicative ((<|>))
+import Data.Data (Data)
+import Data.Semigroup
+import Prelude hiding (lookup)
+
+import qualified Data.Map
+import qualified Data.Set
+
+{-| A `Map` that remembers the original ordering of keys
+
+    This is primarily used so that formatting preserves field order
+
+    This is done primarily to avoid a dependency on @insert-ordered-containers@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
+    deriving (Data, Eq, Foldable, Functor, Show, Traversable)
+
+instance Ord k => Data.Semigroup.Semigroup (Map k v) where
+    (<>) = union
```

Comment:
Might I suggest strategically avoiding defining `Semigroup` and `Monoid` for `Map`? These forgetful instances often do the wrong thing for `Data.Map.Map`; if you want the left-biased behaviour, it's often best to write `union` explicitly.

Reply:
Oh, now I remember why I added it.  `Dhall.RecordType` depends on `Map` being a `Monoid`.  I could add a newtype wrapper there just for this purpose, though